Gossipsub v1.1

AgeManning opened the pull request:
# Description

This PR upgrades the current gossipsub implementation to support the [v1.1 spec](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md). 

It adds a number of features, bug fixes and performance improvements. 

Besides support for all new 1.1 features, other improvements that are of particular note: 

- Improved duplicate LRU-time cache (this was previously a severe bottleneck for large message throughput topics)
- Extended message validation configuration options
- Arbitrary topics (users can now implement their own hashing schemes)
- Improved message validation handling - Invalid messages are no longer dropped but sent to the behaviour for application-level processing (including scoring)
- Support for floodsub, gossipsub v1 and gossipsub v2 
- Protobuf encoding has been shifted into the behaviour. This has permitted two improvements:
     1. Message size verification during publishing (report to the user if the message is too large before attempting to send).
     2. Message fragmentation. If an RPC is too large it is fragmented into its sub components and sent in smaller chunks. 

### Additional Notes

The peer eXchange protocol defined in the v1.1 spec is inactive in its current form. The current implementation permits sending `PeerId` in `PRUNE` messages, however a `PeerId` is not sufficient to form a new connection to a peer. A `Signed Address Record` is required to safely transmit peer identity information. Once these are confirmed (https://github.com/libp2p/specs/pull/217) a future PR will implement these and make PX usable. 


##### Final Note

This is currently undergoing further testing and an audit and I may modify this slightly as bugs come up :)

AgeManning committed:
Gossipsub message signing

AgeManning committed:
Update ipfs-private example

AgeManning committed:
Add optimisation to prevent key calculation each message

AgeManning committed:
Support multiple versions

AgeManning committed:
Initial porting of peer_scoring from go-libp2p

AgeManning committed:
Handle forwarding of signed messages and reviewers comments

AgeManning committed:
Merge latest gossipsub signing

AgeManning committed:
Progress to porting scoring functions

AgeManning committed:
Merge branch 'master' into gossipsub-signing

AgeManning committed:
Finish first port of score params

AgeManning committed:
Add defaults to scoring

AgeManning committed:
Generalise topic hashing

AgeManning committed:
Update topic hashing structure

AgeManning committed:
Update v1.1 tests

AgeManning committed:
Add more unit tests

mxinden committed:
protocols/gossipsub/src/protocol: Use quickcheck to test signing (#31)

AgeManning committed:
Shift signing into behaviour

AgeManning committed:
Merge latest master

AgeManning committed:
Address reviewers suggestions

AgeManning committed:
Send subscriptions to all peers

AgeManning committed:
Update examples

AgeManning committed:
Shift signing option into the behaviour

AgeManning committed:
Revert changes to core/

AgeManning committed:
Reviewers suggestion

AgeManning committed:
Merge remote-tracking branch 'origin/master' into gossipsub-signing

AgeManning committed:
Merge latest master

AgeManning committed:
Revert changes to core/

rklaehn committed:
Switch gossipsub state to use sets instead of vectors

rklaehn committed:
Make tests pass

rklaehn committed:
Some clippy

AgeManning committed:
Prevent duplicate finding during JOIN

AgeManning committed:
Merge #1675

AgeManning committed:
Update tests, improve logging

AgeManning committed:
Update the default d_lo constant

AgeManning committed:
Handle errors correctly

AgeManning committed:
Add privacy and validation options in the config

blacktemplar committed:
Merge branch 'master' into gossipsub-v1.1

rklaehn committed:
A few improvements (#32)

* Make the GossipsubRpc debug instance a bit nicer

Basically just don't show empty components to reduce visual noise

* Introduce send_message helper function to avoid repetition of boilerplate

This also serves to have a central place to log or otherwise intercept all
outgoing messages to other peers of the behaviour.

* Simplify the arc removal in poll

AgeManning committed:
Improve signing validation logic

rklaehn committed:
Change the gossipsub rpc protocol to use bytes for message ids (#34)

And do the corresponding changes to make this work.

AgeManning committed:
Add optional privacy and validation settings

AgeManning committed:
Correct doc link

AgeManning committed:
Prevent invalid messages from being gossiped

AgeManning committed:
Remove unvalidated messages from gossip. Reintroduce duplicate cache

AgeManning committed:
Send grafts when subscriptions are added to the mesh

AgeManning committed:
Merge latest master

AgeManning committed:
Merge gossipsub-signing

blacktemplar committed:
implement explicit peers

blacktemplar committed:
Merge branch 'gossipsub-v1.1' into gossipsub-v1.1-explicit-peers

# Conflicts:
#	examples/gossipsub-chat.rs
#	examples/ipfs-private.rs
#	protocols/gossipsub/src/behaviour.rs
#	protocols/gossipsub/src/behaviour/tests.rs

AgeManning committed:
Fix ipfs-kad example

blacktemplar committed:
Merge branch 'gossipsub-v1.1' into gossipsub-v1.1-explicit-peers

blacktemplar committed:
finish merging + rustfmt

blacktemplar committed:
undo changes in ipfs-private

AgeManning committed:
Only add messages to memcache if not duplicates

AgeManning committed:
Merge latest signing

blacktemplar committed:
Merge pull request #36 from blacktemplar/gossipsub-v1.1-explicit-peers

Gossipsub v1.1 explicit peers

AgeManning committed:
Add mesh maintenance tests and remove excess peers from mesh

AgeManning committed:
Apply reviewers suggestions

AgeManning committed:
Merge latest master

AgeManning committed:
Merge latest gossipsub-signing

AgeManning committed:
Wrap comments

blacktemplar committed:
implement px and prune backoff + simplify tests for explicit peers

blacktemplar committed:
Merge pull request #37 from sigp/gossipsub-v1.1-prune-backoff-and-px

implement PX and prune backoff + simplify tests for explicit peers

blacktemplar committed:
improved data structure for storing backoffs + adding configurable slack for grafting to peers after backoff

blacktemplar committed:
Merge pull request #38 from sigp/gossipsub-v1.1-prune-backoff-and-px

improved data structure for storing backoffs + slack backoff

blacktemplar committed:
allow flood publishing (on by default)

blacktemplar committed:
Merge pull request #39 from sigp/gossipsub-v1.1-flood-publishing

allow flood publishing (on by default)

blacktemplar committed:
implement adaptive gossip dissemination + complete config debug output

blacktemplar committed:
Merge pull request #40 from sigp/gossipsub-v1.1-adaptive-gossip-dissemnation

implement adaptive gossip dissemination + complete config debug output

blacktemplar committed:
fix bug in backoff storage data structure

blacktemplar committed:
make GossipsubConfig fields private + add getters + improve validation in builder

blacktemplar committed:
implements a configurable minimum number of mesh peers that are outbound

blacktemplar committed:
Merge pull request #41 from sigp/gossipsub-v1.1-outbound-mesh-quota

implements a configurable minimum number of mesh peers that are outbound

AgeManning committed:
Ensure sequence number is sent

AgeManning committed:
Merge latest master

AgeManning committed:
Maintain the debug trait

blacktemplar committed:
use the score module in the behaviour, calling the score update method hooks + using the score of peers in the behaviour logic

blacktemplar committed:
Merge pull request #42 from sigp/gossipsub-v1.1-score-wiring

use the score module in the behaviour

blacktemplar committed:
implement opportunistic grafting

blacktemplar committed:
remove done TODO

blacktemplar committed:
remove done TODO

blacktemplar committed:
cargo fmt

blacktemplar committed:
Merge branch 'gossipsub-v1.1' into gossip-v1.1-opportunistic-grafting

blacktemplar committed:
ignore grafts in unknown topics

blacktemplar committed:
Merge pull request #43 from sigp/gossip-v1.1-opportunistic-grafting

Gossip v1.1 opportunistic grafting

blacktemplar committed:
Merge branch 'gossipsub-v1.1' into gossipsub-v1.1-spam-protection-measures

blacktemplar committed:
cap the number of gossipsub retransmission for the iwants of the same message for the same peer

blacktemplar committed:
cap the number of processed ihaves per peer per heartbeat, plus the maximum number of processed message ids inside ihaves per peer per heartbeat + cap the number of message ids we send inside an ihave

AgeManning committed:
Replace lru cache with custom duplicatecache

blacktemplar committed:
track iwant messages we send and penalize peers not responding to them

blacktemplar committed:
removes todo

AgeManning committed:
Improve stream management and error handling

blacktemplar committed:
Merge pull request #44 from sigp/gossipsub-v1.1-spam-protection-measures

Implement missing spam protection measures

blacktemplar committed:
Merge remote-tracking branch 'origin/gossipsub-signing' into gossipsub-v1.1

# Conflicts:
#	protocols/gossipsub/src/behaviour.rs
#	protocols/gossipsub/src/config.rs
#	protocols/gossipsub/src/mcache.rs
#	protocols/gossipsub/src/protocol.rs

blacktemplar committed:
Merge branch 'master' into gossipsub-v1.1

# Conflicts:
#	examples/gossipsub-chat.rs
#	examples/ipfs-private.rs
#	protocols/gossipsub/src/behaviour.rs
#	protocols/gossipsub/src/behaviour/tests.rs
#	protocols/gossipsub/src/config.rs
#	protocols/gossipsub/src/handler.rs
#	protocols/gossipsub/src/lib.rs
#	protocols/gossipsub/src/mcache.rs
#	protocols/gossipsub/src/protocol.rs
#	protocols/gossipsub/src/topic.rs
#	protocols/gossipsub/tests/smoke.rs

blacktemplar committed:
cargo fmt

blacktemplar committed:
Gossipsub v1.1 check fmt (#45)

add workflow that checks fmt for all branches and pull requests in package gossipsub

AgeManning committed:
Merge duplicate cache

AgeManning committed:
Merge substream updates

AgeManning committed:
Propagate signature verification and peer kinds to behaviour (#46)

* Propagate signature verification and peer kinds to behaviour

* Propagate validation errors to the behaviour

* handle peer kind + fix flaky test

* Inform the behaviour of invalid messages

* Correct duplicate cache

* penalize invalid messages

* generalize DuplicateCache to TimeCache + implement minimalistic Entry API + use it for deliveries tracking in scoring module

* fix warnings in test

* remove accidentally added benchmark functions

* Handle peers that do not support the protocol

Co-authored-by: blacktemplar <blacktemplar@a1.net>

blacktemplar committed:
treat peers that do not support the protocol as disconnected (#48)

blacktemplar committed:
Gossipsub v1.1 various improvements (#49)

* remove dbg! calls and add debug logging for peer scoring

* export MessageAcceptance and rename validate_message to report_message_validation_result to also signal that this message should get called in case of invalid messages

* fix double reject_message call

* gossip promises are fulfilled already on receiving the message without validation

* derive debug for MessageAcceptance

* add helper method to get config builder from existing config

* allow adding/changing TopicScoreParams during runtime

* more debug output for messages from self

* fixes incompatibility with anonymous PeerId in lighthouse

* cargo fmt

* more debug output for broken promises

AgeManning committed:
Gossipsub 1.1 ammendments (#50)

* Shuffling things around

* Further minor ammendments

* Misc typos and improvements

* Correct tests

* Wasm support

AgeManning committed:
Complete porting of go scoring tests (#51)

* Update tests

* Complete scoring tests

AgeManning committed:
Add peer blacklisting (#52)

AgeManning committed:
Estimate message size (#53)

AgeManning committed:
Message fragmentation (#54)

* Estimate message size

* Fragment RPC messages

* Shift protobuf encoding into behaviour

AgeManning committed:
Correct tests and merge latest master (#55)

AgeManning committed:
Merge latest master

AgeManning committed:
Remove github workflow

AgeManning committed:
Cleanup code add logs (#56)

* Clean up code

* Upgrade crates

* Update example to use signing

* Add debug log

* Improve keepalive for peers without gossipsub support

* Improve debug logs for new peer types

AgeManning committed:
Further improve logging (#57)

* Improve logging

* Print invalid seq_no length

* Improve debug logs

AgeManning committed:
Remove lru_time_cache_dep (#58)

mxinden reviewed (commented):
Just a couple of small nits. I need to spend more time to fully understand Gossipsub v1.1 to give a thorough review.

mxinden reviewed (commented):
Still only small nits that caught my attention while understanding the greater picture.

blacktemplar committed:
Various improvements (#60)

* add the new behaviour penalty threshold parameter

* fix bug to request message ids in IWANT only if they are not in duplication cache

* more public methods for inspecting the state, distinguish between all mesh peers and all peers functions

* return message ids of published messages

* contains methods for time_cache

* interpret empty bytes the same as missing optional protobuf field

* fix missing field resets when setting mesh state to inactive

* add optional message delivery time callbacks

* handle updating gossipsub parameters for decreasing caps

* improved parameter setting api

* fix wrong order of greylist checking

* more debug and tracing logging

* fix tests

blacktemplar committed:
Cache published message ids (#61)

* add a cache for published message ids to detect self published messages

* more debug+trace logs

* cargo fmt

blacktemplar committed:
use generic type for message data + cache message ids (#67)

* use generic type for message data + cache message ids

* fix docs

* fix examples

* fix examples

blacktemplar committed:
Adds a fast message id function and an additional cache for it (#68)

* add a fast message id function and an additional cache for it

* also test number of calls of `From<Vec<u8>>` messsage data transformations

* Add documentation for fast message id function. Change fast message id function return type to avoid confusing message ids with fast message ids.

blacktemplar committed:
Merge branch 'master' into gossipsub-v1.1

# Conflicts:
#	protocols/gossipsub/Cargo.toml
#	protocols/gossipsub/src/behaviour.rs
#	protocols/gossipsub/tests/smoke.rs

blacktemplar committed:
export FastMessageId

blacktemplar committed:
Gossipsub v1.1 bug fixes (#71)

* Allow helper functions - remove warnings

* don't graft to negative scored peers after subscription

* improve backoff handling

Co-authored-by: Age Manning <Age@AgeManning.com>

blacktemplar committed:
add subscription filters + some useful default filters + refactor tests by using builder pattern (#70)

* add subscription filters + some useful default filters

* cargo fmt + log when incoming subscriptions get filtered + variable renaming

blacktemplar committed:
remove support for multi-topic messages (#73)

* remove support for multi-topic messages

* make message topic required

blacktemplar committed:
also consider allow self origin when using a published message ids dict (#84)

* also consider allow self origin when using a published message ids dictionary

* make allow_self_origin configurable via the config builder

blacktemplar committed:
fix clippy warnings (#83)

blacktemplar committed:
Merge branch 'master' into gossipsub-v1.1

# Conflicts:
#	protocols/gossipsub/Cargo.toml

AgeManning committed:
Merge latest master

AgeManning committed:
Merge latest master

AgeManning committed:
Address reviewers comments (#88)

* Address reviewers comments

* fix rust doc linking error

* Add copy trait

Co-authored-by: blacktemplar <blacktemplar@a1.net>

blacktemplar committed:
implement as_ref and as_static_ref to get a &str from PeerKind (#91)

AgeManning committed:
Merge latest master

AgeManning committed:
Update to latest master

blacktemplar committed:
only respond to IWANT if message got already validated (#96)

mxinden committed:
protocols/gossipsub: Review (#93)

* protocols/gossipsub: Review

* Addressing most of the review comments

* Address reviewers comments

* Appease CI

Co-authored-by: blacktemplar <blacktemplar@a1.net>
Co-authored-by: Age Manning <Age@AgeManning.com>

AgeManning committed:
Merge latest master

AgeManning committed:
Add optional compression to gossipsub (#100)

* Update rustls requirement from 0.18.0 to 0.19.0 (#1852)

Updates the requirements on [rustls](https://github.com/ctz/rustls) to permit the latest version.
- [Release notes](https://github.com/ctz/rustls/releases)
- [Changelog](https://github.com/ctz/rustls/blob/main/OLDCHANGES.md)
- [Commits](https://github.com/ctz/rustls/compare/v/0.18.0...v/0.19.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Prepare libp2p-websocket-0.26.1

* Update top-level libp2p-websocket patch version.

* [request-response] Refine success & error reporting for inbound requests. (#1867)

* Refine error reporting for inbound request handling.

At the moment one can neither get confirmation when a
response has been sent on the underlying transport, nor
is one aware of response omissions. The latter was
originally intended as a feature for support of
one-way protocols, which seems like a bad idea in
hindsight. The lack of notification for sent
responses may prohibit implementation of some
request-response protocols that need to ensure
a happens-before relation between sending a
response and a subsequent request, besides uses
for collecting statistics.

Even with these changes, there is no active notification
for failed inbound requests as a result of connections
unexpectedly closing, as is the case for outbound requests.
Instead, for pending inbound requests this scenario
can be identified if necessary by the absense of both
`InboundFailure` and `ResponseSent` events for a particular
previously received request. Interest in this situation is
not expected to be common and would otherwise require
explicitly tracking all inbound requests in the `RequestResponse`
behaviour, which would be a pity. `RequestResponse::send_response`
now also synchronously returns an error if the inbound upgrade
handling the request has been aborted, due to timeout or
closing of the connection, giving more options for graceful
error handling for inbound requests.

As an aside, the `Throttled` wrapper now no longer emits
inbound or outbound error events occurring in the context
of sending credit requests or responses. This is in addition
to not emitting `ResponseSent` events for ACK responses of
credit grants.

* Update protocols/request-response/src/lib.rs

Co-authored-by: Max Inden <mail@max-inden.de>

* Address some minor clippy warnings. (#1868)

* Track pending credit request IDs.

In order to avoid emitting events relating to credit grants or acks
on the public API. The public API should only emit events relating
to the actual requests and responses sent by client code.

* Small cleanup

* Cleanup

* Update versions and changelogs.

* Unreleased

Co-authored-by: Max Inden <mail@max-inden.de>

* core/benches: Add rudimentary benchmark for PeerId::from_bytes and clone (#1875)

* core: Add rudimentary benchmark for PeerId::from_bytes and clone

* .github/workflow: Include benchmarks

To ensure changes through pull requests won't make benchmarks fail to
compile or run, run them as part of CI.

* [mdns] Split response packets if necessary. (#1877)

* [mdns] Split response packets.

Prevent MDNS response packets becoming too large by creating
multi-packet responses. Also skip addresses that don't fit
into a TXT record or contain invalid characters.

* Update protocols/mdns/src/dns.rs

Co-authored-by: Max Inden <mail@max-inden.de>

* Refactor response packet construction.

* Update mdns changelog.

Co-authored-by: Max Inden <mail@max-inden.de>

* core/benches: Add PeerId sort_vec benchmark (#1878)

* Prepare v0.32 (#1879)

* [websocket] Update minimum async-tls patch version. (#1881)

* Update minimum async-tls patch version.

After the upgrade to rustls 0.19, this is the minimum
version required to build.

* Prepare libp2p patch.

* Update async-tls requirement from 0.10.2 to 0.11.0 (#1884)

* Update async-tls requirement from 0.10.2 to 0.11.0

Updates the requirements on [async-tls](https://github.com/async-std/async-tls) to permit the latest version.
- [Release notes](https://github.com/async-std/async-tls/releases)
- [Commits](https://github.com/async-std/async-tls/compare/v0.10.2...v0.11.0)

Signed-off-by: dependabot[bot] <support@github.com>

* *: Prepare release

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Max Inden <mail@max-inden.de>

* rename generic types + add default types + update documentation links

* fix broken intra links

* fix tests

* type annotation in example

* fix doc example

* Add compression to gossipsub

* Add snappy compression to the smoke tests

* Stack allocated PeerId (#1874)

* Stack allocate PeerId.

* Update stuff.

* Upgrade rusttls to fix build.

* Remove unnecessary manual implementations.

* Remove PeerId::into_bytes.

* Remove bytes dependency.

* Perform some cleanup.

* Use Into<kbucket::Key<K>>.

* Update versions and changelogs.

* Fix PR link.

* Fix benchmarks.

Co-authored-by: Roman S. Borschel <roman@parity.io>

* Remove copy and adjust validation ordering

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Roman S. Borschel <roman@parity.io>
Co-authored-by: Roman Borschel <romanb@users.noreply.github.com>
Co-authored-by: Max Inden <mail@max-inden.de>
Co-authored-by: blacktemplar <blacktemplar@a1.net>
Co-authored-by: David Craven <david@craven.ch>

AgeManning committed:
Remove feature gating from gossipsub

AgeManning committed:
Merge latest master

AgeManning committed:
Remove compression in favour of a more general data transform

AgeManning committed:
Loosen trait bounds on gossipsub

mxinden reviewed (approved):
I would be in favor of merging this pull request into `master`. As far as I can tell, the only outstanding task would be to fix the intra-doc links. What do you think @AgeManning @blacktemplar?

AgeManning committed:
Merge latest master

AgeManning committed:
Fix doclinks

diegomrsantos commented on protocols/gossipsub/src/peer_score/params.rs:
```diff
@@ -0,0 +1,399 @@
+impl Default for PeerScoreThresholds {
+    fn default() -> Self {
+        PeerScoreThresholds {
+            gossip_threshold: -10.0,
+            publish_threshold: -50.0,
+            graylist_threshold: -80.0,
+            accept_px_threshold: 10.0,
+            opportunistic_graft_threshold: 20.0,
```
@AgeManning could you please share how those values were defined?

AgeManning replied:
These essentially can be defined for whatever network you are running gossipsub on. Its been a few years, but I believe these were from the go-implementation at the time, which may have been used in some of the tests in the v1.1 gossipsub paper where they tested a few attack vectors. 

These numbers shouldn't be set independently, but based on the maximum and minimum values you can get from the score.

diegomrsantos replied:
I found this PR that says the values were inspired by LH https://github.com/status-im/nimbus-eth2/pull/2091. These are the values used:

```
     gossipThreshold: -4000,
     publishThreshold: -8000,
     graylistThreshold: -16000
```

They are used here indeed https://gist.github.com/blacktemplar/5c1862cb3f0e32a1a7fb0b25e79e6e2c#file-51000-toml-L33, but LH doesn't use to configure GossipSub, but rather here https://github.com/libp2p/rust-libp2p/blob/25d14349cc32f7a8e3feffb931ba2e84cccb0013/protocols/gossipsub/src/metrics.rs#L76

AgeManning replied:
Now I"m confused about what you are asking. 

The values in this repository i.e `gossip_threshold: -10` is not the same value as what we have chosen in the Ethereum consensus network i.e `-4000`. 

As I mentioned, these values need to be defined for whatever network you are running. For Ethereum's consensus network, we did a bunch of estimations and determined a bunch of numbers. That is entirely independent of the `gossipsub` protocol, which should run on any network. 

If you question is specific to lighthouse or the Ethereum network, then that is a different question and probably should be asked on the lighthouse repo as those values are for a specific network. 

You did point out that the metrics value uses the Ethereum consensus values as default. This probably should be using the same defaults as the rust-libp2p repo. At the time, Lighthouse was the only one consuming these metrics and so was built around our scoring values to get sane results in the histograms.

diegomrsantos replied:
You're right, sorry for the confusion. It's clear now, thanks. I somehow misled myself looking at this PR that introduces params in the lighthouse rust-libp2p fork https://github.com/sigp/rust-libp2p/pull/556/files#diff-9f622a0085d0d96aef9e883b7afdcdd4a652e8801be7cfabf82819f5a74a70e0. Is there a lighthouse PR defining the values for lighthouse?

AgeManning replied:
What specifically are you interested in?

Are you asking how we figured out what values to use in Lighthouse? There is an external document I can share with you if that's the case. The PR to lighthouse will likely just show values being added with no context. 

Feel free to DM me in the Lighthouse discord if you like
